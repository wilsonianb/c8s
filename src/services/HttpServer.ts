import * as Hapi from '@hapi/hapi'
import registerContainersController from '../controllers/containers'
import registerStaticController from '../controllers/static'
import registerProxyController from '../controllers/proxy'
import SelfTestCheck from '../services/SelfTestCheck'
import { Injector } from 'reduct'
import * as Inert from '@hapi/inert'
import * as Vision from '@hapi/vision'
import * as Handlebars from 'handlebars'
import * as path from 'path'
import Config from './Config'

import { create as createLogger } from '../common/log'
const log = createLogger('HttpServer')

export default class HttpServer {
  private config: Config
  private selfTestCheck: SelfTestCheck
  private server: Hapi.Server

  constructor (deps: Injector) {
    this.config = deps(Config)
    this.selfTestCheck = deps(SelfTestCheck)
    this.server = new Hapi.Server({
      uri: this.config.publicUri.replace(/\/+$/, ''),
      address: this.config.bindIp,
      port: this.config.port,
      mime: {
        override: {
          // Streaming responses shouldn't get buffered so for simplicity we'll
          // just turn off compression for them.
          'application/vnd.codius.raw-stream': {
            compressible: false
          }
        }
      }
    })

    registerContainersController(this.server, deps)
    registerStaticController(this.server, deps)
    registerProxyController(this.server, deps)
  }

  async init () {
    await this.server.register({ plugin: require('@hapi/h2o2') })
    await this.server.register(this.selfTestCheck.checkSelfTestPlugin)
    await this.server.register(Inert)
    await this.server.register(Vision)
    this.server.views({
      engines: {
        html: Handlebars
      },
      relativeTo: path.resolve(__dirname, '../'),
      path: 'public/templates'
    })

    this.server.route({
      method: 'GET',
      path: '/assets/app.bundle.js',
      handler: {
        file: path.join(__dirname, '../public/assets/app.bundle.js')
      }
    })

    this.server.route({
      method: 'GET',
      path: '/assets/proxy.bundle.js',
      handler: {
        file: path.join(__dirname, '../public/assets/proxy.bundle.js')
      }
    })
    await this.server.initialize()
  }

  async start () {
    await this.server.start()

    log.info('listening at %s', this.server.info.uri)
  }

  async stop () {
    await this.server.stop()

    log.info('server stopped')
  }

  getServer () {
    if (process.env.NODE_ENV === 'test') {
      return this.server
    }
    return null
  }
}
