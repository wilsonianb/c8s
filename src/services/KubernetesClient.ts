import { Client1_13 } from 'kubernetes-client'
const Request = require('kubernetes-client/backends/request')
import { PodSpec } from '../schemas/PodSpec'
import * as Boom from 'boom'
import Config from './Config'
import { Injector } from 'reduct'
import { IncomingMessage } from 'http'

import { create as createLogger } from '../common/log'
const log = createLogger('KubernetesClient')

export interface PodInfoResponse {
  kind: string
  apiVersion: string
  metadata: {
    name: string,
    namespace: string,
    selfLink: string,
    uid: string,
    resourceVersion: string,
    creationTimestamp: string
  },
  spec: {
    containers: {
      name: string,
      image: string,
      env: {
        name: string
        value: string
      }[],
      resources: {
        limits: {
          cpu: string,
          memory: string
        },
        requests: {
          cpu: string,
          memory: string
        }
      },
      terminationMessagePath: string,
      terminationMessagePolicy: string,
      imagePullPolicy: string
    }[],
    restartPolicy: string,
    terminationGracePeriodSeconds: string,
    dnsPolicy: string,
    serviceAccountName: string,
    serviceAccount: string,
    automountServiceAccountToken: boolean,
    nodeName: string,
    securityContext: object,
    schedulerName: string,
    tolerations: {
      key: string,
      operator: string,
      effect: string,
      tolerationSeconds: number
    }[],
    priority: number,
    runtimeClassName: string,
    enableServiceLinks: boolean
  },
  status: {
    phase: string,
    conditions: {
      type: string,
      status: string,
      lastProbeTime: string,
      lastTransitionTime: string
    }[],
    hostIP: string,
    podIP: string,
    startTime: string,
    containerStatuses: {
      name: string,
      state: {
        running: {
          startedAt: string
        }
      },
      lastState: {},
      ready: boolean,
      restartCount: number,
      image: string,
      imageID: string,
      containerID: string
    }[],
    qosClass: string
  }
}

export default class KubernetesClient {
  private client: any
  private config: Config

  constructor (deps: Injector) {
    this.config = deps(Config)

    const devConfig = {
      auth: {
        bearer: 'token'
      },
      ca: 'ca',
      namespace: 'namespace',
      url: 'https://host:443'
    }

    const k8sConfig = this.config.devMode ? devConfig : Request.config.getInCluster()
    this.client = new Client1_13({ backend: new Request(k8sConfig) })
  }

  async getPodInfo (podId: string): Promise<PodInfoResponse> {
    log.debug(`fetching pod info. id=${podId}`)
    const response = await this.client.api.v1.namespaces(this.config.k8sNamespace).pods(podId).get()
    return response.body
  }

  async getPodIP (hash: string): Promise<string> {
    const info = await this.getPodInfo(hash)
    return info.status.podIP
  }

  async getPodList (): Promise<Array<string>> {
    const res = await this.client.api.v1.namespaces(this.config.k8sNamespace).pods.get({
      qs: {
        includeUninitialized: true
      }
    })

    const pods = res.body.items.reduce((acc: Array<string>, pod: PodInfoResponse) => {
      if (pod.status.phase === 'Running') {
        return [...acc, pod.metadata.name]
      }
      return acc
    }, [])

    log.debug(`running pod list=${pods}`)
    return pods
  }

  async deletePod (podId: string): Promise<void> {
    log.info('deleting pod. id=%s', podId)
    try {
      await this.client.api.v1.namespaces(this.config.k8sNamespace).pods(podId).delete()
    } catch (e) {
      throw Boom.serverUnavailable('Could not delete pod: error=' + e.message)
    }
  }

  // rename create
  async runPod (podSpec: PodSpec): Promise<void> {
    log.info('creating pod. id=%s', podSpec.metadata.name)
    try {
      await this.client.api.v1.namespaces(this.config.k8sNamespace).pods.post({ body: podSpec })
      await new Promise(async (resolve, reject) => {
        const stream = await this.client.api.v1.watch.namespaces(this.config.k8sNamespace).pods(podSpec.metadata.name).getObjectStream()

        const timer = setTimeout(() => {
          reject(new Error('Timed out before pod reached Ready state'))
        }, 60000)

        stream.on('data', (pod: any) => {
          if (pod.object && pod.object.status.phase === 'Running') {
            clearTimeout(timer)
            resolve()
          }
        })
      })
    } catch (e) {
      throw Boom.serverUnavailable('Could not create pod: error=' + e.message)
    }
  }

  getLog (podId: string, containerId: string, follow: boolean = false): Promise<IncomingMessage> {
    log.info('attaching to pod. id=%s', podId)
    try {
      return this.client.api.v1.namespaces(this.config.k8sNamespace).pods(podId).log.getByteStream({
        qs: {
          container: containerId,
          follow: true
        }
      })
    } catch (e) {
      throw Boom.serverUnavailable('failed to attach to pod: error=' + e.message)
    }
  }
}
