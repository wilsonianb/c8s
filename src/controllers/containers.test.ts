import HttpServer from '../services/HttpServer'
import KubernetesClient from '../services/KubernetesClient'
import reduct from 'reduct'
import { ServerInjectResponse } from '@hapi/hapi'
const deps = reduct()
const server = deps(HttpServer)
const k8sClient = deps(KubernetesClient)

describe('containers controller', () => {

  const options = {
    headers: {
      Accept: `application/codius-v1.0.0+json`,
      'Content-Type': 'application/json',
      Host: 'localhost'
    },
    method: 'POST',
    url: '/containers'
  }

  beforeAll(async () => {
    await server.init()
  })

  afterAll(async () => {
    await server.stop()
  })

  test('accepts POST request with valid manifest', async () => {
    const k8sCreateSpy = jest.spyOn(k8sClient, 'createKnativeService')
    k8sCreateSpy.mockImplementation(() => Promise.resolve())

    const response: ServerInjectResponse = await server.getServer()!.inject({
      ...options,
      payload: JSON.stringify({
        manifest: {
          name: 'test-container',
          version: '1.0.0',
          machine: 'small',
          port: '8080',
          containers: [{
            id: 'c8s-container',
            image: 'androswong418/codius-test-image@sha256:0dce885c4e558a8a7612b80e3c7f5faa54520ed27836b1f892962cf855d031a2',
            environment: {}
          }],
          vars: {}
        },
        private: {
          vars: {}
        }
      })
    })

    expect(response.statusCode).toBe(200)
    expect(k8sCreateSpy).toHaveBeenCalled()
    expect(response.result).toMatchSnapshot()
  })

  test('rejects non-JSON payload', async () => {
    const response: ServerInjectResponse = await server.getServer()!.inject({
      ...options,
      payload: 'not JSON'
    })

    expect(response.statusCode).toBe(400)
    if (response.result) {
      expect(response.result['message']).toBe('Invalid request payload JSON format')
    } else {
      fail('response.result not defined')
    }
  })

  test('rejects invalid manifest payload', async () => {
    const response: ServerInjectResponse = await server.getServer()!.inject({
      ...options,
      payload: JSON.stringify({
        this: 'is not a valid manifest'
      })
    })

    expect(response.statusCode).toBe(400)
    if (response.result) {
      expect(response.result['message']).toBe('Invalid request payload input')
    } else {
      fail('response.result not defined')
    }
  })
})
