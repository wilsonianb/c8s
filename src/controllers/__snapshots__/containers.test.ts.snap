// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers controller accepts POST request with valid manifest 1`] = `
Object {
  "manifestHash": "73vneknj64kajk3f5xcf5ckknvpbwihjcal7bff7sj26y3pnnjha",
  "url": "http://73vneknj64kajk3f5xcf5ckknvpbwihjcal7bff7sj26y3pnnjha.local.codius.org:3000/",
}
`;
